packages/product/src/services/product-module-service.ts

```diff
@@ -563,15 +578,100 @@ export default class ProductModuleService<
       }))
     )
 
-    return createdProducts
+    return Array.isArray(data) ? createdProducts : createdProducts[0]
   }
 
+  async upsert(
+    data: ProductTypes.UpsertProductDTO[],
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductDTO[]>
+  async upsert(
+    data: ProductTypes.UpsertProductDTO,
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductDTO>
+  @InjectTransactionManager("baseRepository_")
+  async upsert(
+    data: ProductTypes.UpsertProductDTO[] | ProductTypes.UpsertProductDTO,
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<ProductTypes.ProductDTO[] | ProductTypes.ProductDTO> {
+    const input = Array.isArray(data) ? data : [data]
+    const forUpdate = input.filter(
+      (product): product is UpdateProductInput => !!product.id
+    )
+    const forCreate = input.filter(
+      (product): product is ProductTypes.CreateProductDTO => !product.id
+    )
+
+    let created: Product[] = []
+    let updated: Product[] = []
+
+    if (forCreate.length) {
+      created = await this.create_(forCreate, sharedContext)
+    }
+    if (forUpdate.length) {
+      updated = await this.update_(forUpdate, sharedContext)
+    }
+
+    const result = [...created, ...updated]
```

Comment:
thought(non-blocking): could these be parallelized?
```typescript
    if (forCreate.length) {
      created = this.create_(forCreate, sharedContext)
    }
    if (forUpdate.length) {
      updated = this.update_(forUpdate, sharedContext)
    }

    const result = await Promise.all([created, updated]).then(res => res.flat())
```

Reply:
@srindom I didn't do it with flat as I want to distingush between what updated and what was created for the events. It can be done without flattening though, but I figured in reality it is rare to do both create and update at the same time, so I figured no need to add that for now.

Reply:
I see - all good 👍